### ISSUE TITLE
Using a built executable as compiler

### ISSUE THREAD

**Issue by Leandros** [fastbuild/fastbuild#162]
I'm trying to use an executable, which is build as part of the project, as a compiler for other objects in the project.

FASTBuild, unfortunately, refuses this:

```
BFF file 'C:\perforce\thor\res\models\fbuild.bff' has changed (reparsing will occur).
C:\perforce\thor\res\models\fbuild.bff(22,1): FASTBuild Error #1102 - ObjectList() - 'Compiler' ('C:\Users\Arvid\workspace\perforce\thor\_out\bin\objbin.ex e') is of unexpected type 'Exe'. Expected 'Compiler'.
ObjectList('res-models-obj')
^
\--here
```

Said executable is compiled via `Executable` and I want to use it as `.Compiler` in `ObjectList`.

Is there any way to make this work?

**Comment by ffulin** [comment]
FASTBuild expects the .Compiler to point something declared with Compiler(), but for convenience it automatically upgrades a single file to a Compiler() declaration for you.  If you provide anything more complex, it can't do this automatic conversion, and so you get this error.

It should be possible to make this work as follows:
```C++
Executable( 'CustomExecutable' ) // Your executable
{
    ...

Compiler( 'CustomCompiler' )
{
    .Executable = 'CustomExecutable'
    ...

.Objectlist( 'ObjectList' )
{
    .Compiler = 'CustomCompiler'
```

**Comment by Leandros** [comment]
That was also one of my first ideas to try, sadly FASTBuild complains the compiler I try to define has already been defined:

```
BFF file 'C:\Users\Arvid\workspace\perforce\thor\res\models\fbuild.bff' has changed (reparsing will occur).
C:\Users\Arvid\workspace\perforce\thor\res\models\fbuild.bff(10,1): FASTBuild Error #1100 - Compiler() - Target 'C:\Users\Arvid\workspace\perforce\thor\_out\bin\objbin.exe' already defined.
Compiler('objbin-compiler')
^
\--here
```

**Comment by ffulin** [comment]
I'm afraid you are right: this won't currently work :(

To make this work, a change is required to Compiler() to allow it to reference the primary executable from another target, instead of requiring it to be the entry that declares that file. This would have the added benefit of allowing multiple Compiler() instances to reference the same primary executable, which would be a useful improvement to the flexibility of FASTBuild.

I'll investigate further to see how easy/difficult this change might be.

**Comment by Leandros** [comment]
Merci. 🙇

**Comment by gkoreman** [comment]
I looked into this a little bit and it gets a bit sticky when trying to change the Compiler node from being a MetaName() and MetaFile() to a MetaNone(). It looks like it can be done by special casing more things (like the GetStamp() functionality) on the compiler. 
@ffulin do you have any thoughts on this?

**Comment by gkoreman** [comment]
Potential fix in #410

**Comment by ffulin** [comment]
The next version (v0.96) will include @gkoreman's changes, allowing for FASTBuild to use Compilers it has compiled itself. I will leave this issue open until that version is released.

**Comment by ffulin** [comment]
A version containing this new feature has now been released (v0.96).
